JDK-8298060: Fix precision bug in gesture recognizer classes

hjohn opened the pull request:
This includes a fix for the precision problem we've found as part of the graphics warnings clean ups.

I've included two commits, one with just the minimal fix, and one with the clean ups. I can drop off the 2nd commit if it is deemed to be of no added value.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298060](https://bugs.openjdk.org/browse/JDK-8298060): Fix precision bug in gesture recognizer classes


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/966/head:pull/966` \
`$ git checkout pull/966`

Update a local copy of the PR: \
`$ git checkout pull/966` \
`$ git pull https://git.openjdk.org/jfx.git pull/966/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 966`

View PR using the GUI difftool: \
`$ git pr show -t 966`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/966.diff">https://git.openjdk.org/jfx/pull/966.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/966#issuecomment-1337284404)

arapte commented:
The commit#1 changes look good.
I would recommend to revert commit#2 for few reasons:
1. Cleanup changes are a lot more than actual fix change.
2. It will be difficult to revisit in future.
3. If required the cleanup changes could go in as a separate fix.

Could you please provide link to the discussion that led to this change?

hjohn commented:
> The commit#1 changes look good. I would recommend to revert commit#2 for few reasons:
> 
> 1. Cleanup changes are a lot more than actual fix change.
> 2. It will be difficult to revisit in future.
> 3. If required the cleanup changes could go in as a separate fix.
> 
> Could you please provide link to the discussion that led to this change?

@arapte the discussion was part of fixing warnings in javafx.graphics.  You can find it here (open first resolved comment): https://github.com/openjdk/jfx/pull/960#pullrequestreview-1197937273

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/GestureRecognizer.java:
```diff
@@ -25,5 +25,9 @@
+import java.util.concurrent.TimeUnit;
+
 interface GestureRecognizer extends GlassTouchEventListener {
+    static final long INITIAL_VELOCITY_THRESHOLD_NANOS = 100L * 1000;
```
100_000L ?

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/RotateGestureRecognizer.java:
```diff
@@ -25,76 +25,72 @@
             String s = System.getProperty("com.sun.javafx.gestures.rotate.threshold");
             if (s != null) {
-                ROTATATION_THRESHOLD = Double.valueOf(s);
+                rotationThresholdDegrees = Double.valueOf(s);
```
should we catch a NumberFormatException and re-throw it with a meaningful text message?  Or the current behavior is ok?

ditto on line 64

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/ScrollGestureRecognizer.java:
```diff
@@ -25,61 +25,61 @@
             s = System.getProperty("com.sun.javafx.gestures.scroll.inertia");
             if (s != null) {
-                SCROLL_INERTIA_ENABLED = Boolean.valueOf(s);
+                scrollInertiaEnabled = Boolean.valueOf(s);
```
same comment about re-throwing an exception (x2)

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/ScrollGestureRecognizer.java:
```diff
@@ -245,15 +242,16 @@ public void notifyEndTouchEvent(long time) {
                         factorX = deltaX / scrollMagnitude;
                         factorY = deltaY / scrollMagnitude;
-                        initialInertiaScrollVelocity = scrollMagnitude / timePassed;
+                        initialInertiaScrollVelocity = scrollMagnitude / nanosPassed * NANOS_TO_SECONDS;
```
is this correct?  shouldn't it be `scrollMagnitude / (nanosPassed * NANOS_TO_SECONDS)`?

hjohn replied:
I think it's correct, unless you have reason to believe why it wouldn't be.  `scrollMagnitude` is a double, and the division and multiplication that follow it will be all done as doubles.

andy-goryachev-oracle replied:
your change is equivalent to 

(scrollMagnitude * NANOS_TO_SECONDS) / nanosPassed

is that correct?

hjohn replied:
That would be equivalent yes.

Perhaps it is because `timePassed` was a `double` before measured in seconds while I'm now doing the calculation with nano seconds?

hjohn replied:
Okay I see what you're getting at, your first suggestion would not be correct, it's supposed to be what you suggest in the 2nd one.  So what I want is:

     (scrollMagnitude / nanosPassed) * NANOS_TO_SECONDS

or:

     (scrollMagnitude * NANOS_TO_SECONDS) / nanosPassed

...but that's the same as what I wrote.

kevinrushforth replied:
But is it correct? The previous code computed a deltaTime in seconds as `((currentNanos - startNanos) / 1e9)`, and then divided the `scrollMagnitude` by that. I think that the equivalent to the old code is:

```
    scrollMagnitude / (nanosPassed * NANOS_TO_SECONDS)
```

hjohn replied:
It turns out it is not correct, the parenthesis are indeed required, not sure how I could have missed it. I'm going to see if I can add a test to this as they're missing to ensure it does the same before/after (aside from precision).

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/RotateGestureRecognizer.java:
```diff
@@ -307,10 +298,11 @@ else if (initialInertiaRotationVelocity < -MAX_INITIAL_VELOCITY)
+                        long nanosPassed = nanos - rotationStartNanos;
+
+                        if (nanosPassed > INITIAL_VELOCITY_THRESHOLD_NANOS) {
+                            initialInertiaRotationVelocity = currentRotation / nanosPassed * NANOS_TO_SECONDS;
```
is this correct?  shouldn't it be `currentRotation / (nanosPassed * NANOS_TO_SECONDS)` ?

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/ZoomGestureRecognizer.java:
```diff
@@ -25,61 +25,59 @@
             s = System.getProperty("com.sun.javafx.gestures.zoom.inertia");
             if (s != null) {
-                ZOOM_INERTIA_ENABLED = Boolean.valueOf(s);
+                zoomInertiaEnabled = Boolean.valueOf(s);
```
... and here

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/ZoomGestureRecognizer.java:
```diff
@@ -278,10 +271,11 @@ public void notifyEndTouchEvent(long time) {
+                        long nanosPassed = nanos - zoomStartNanos;
+
+                        if (nanosPassed > INITIAL_VELOCITY_THRESHOLD_NANOS) {
+                            initialInertiaZoomVelocity = (totalZoomFactor - prevTotalZoomFactor) / nanosPassed * NANOS_TO_SECONDS;
```
`(totalZoomFactor - prevTotalZoomFactor) / (nanosPassed * NANOS_TO_SECONDS);` ?

hjohn replied:
The order is irrelevant, it could as well be `((totalZoomFactor - prevTotalZoomFactor) / nanosPassed) * NANOS_TO_SECONDS` -- I'll let the compiler decide what is more optimal.

hjohn replied:
See other conversation, your suggestion would not be correct as it changes the order of the operations.

kevinrushforth reviewed (commented):
inline

hjohn commented:
@kevinrushforth 

Please advise if I should continue on this road:

I've reverted `RotateGestureRecognizer` to add a test case first.

To make `RotateGestureRecognizer` testable I had to make it public and abstract away the need for `ViewScene` as constructor parameter, see the `Make RotateGestureRecognizer easier to test` commit.  I also added a test case (90% coverage).

If you are okay with these changes (I needed to change `GlassViewEventHandler` and add a small interface), I can do the same for the Zoom & ScrollGestureRecognizers. 

The test I added reveals the precision problem when you uncomment the line: `passTime(Long.MAX_VALUE / 1000 / 1000 - 10000000);`  -- the accuracy problem of converting longs to doubles then breaks the test.  The newer version of the code works (after I fixed the calculation by adding parenthesis).

kevinrushforth commented:
> To make RotateGestureRecognizer testable I had to make it public and abstract away the need for ViewScene as constructor parameter, see the Make RotateGestureRecognizer easier to test commit. I also added a test case (90% coverage).

This seems quite an intrusive change just to support low level testing of this. I can take a closer look next week, but I'm not really sold on the idea.

hjohn commented:
I think this change still has merit.  I also think the change to make the code more testable is a good thing.  A lot of code in JavaFX is hard to test, and when things are hard to test, the testing is often done poorly or not at all -- these classes did not have any tests.
